### PR TITLE
Add 'entry' and 'pre_init' attributes

### PR DESCRIPTION
**Pull request by Disasm**
Implementation is based on [`cortex-m-rt-macros`](https://github.com/rust-embedded/cortex-m-rt/tree/master/macros) code.

This implementation has been changed to make `static mut` unsafe inside entry point and different handlers.

Related: https://github.com/rust-embedded/riscv-rt/pull/20

### CODE REVIEW

**Comment by dvc94ch** [comment]
The docs need updating, they contain a lot of references to ARM.

**Comment by Disasm** [comment]
Yes, but I'd like to do this later along with adding RISC-V-related interrupt processing code. ARM-related macros are not exported in `riscv-rt` so this should not cause any misunderstandings. If you think it necessary, I can add `#[doc(hidden)]` for them.

**Comment by Disasm** [comment]
Updated.

**Comment by Disasm** [comment]
@dvc94ch Anything else I need to fix?

File: macros/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+/// Attribute to declare the entry point of the program
+///
+/// **IMPORTANT**: This attribute must appear exactly *once* in the dependency graph. Also, if you
+/// are using Rust 1.30 the attribute must be used on a reachable item (i.e. there must be no
+/// private modules between the item and the root of the crate); if the item is in the root of the
+/// crate you'll be fine. This reachability restriction doesn't apply to Rust 1.31 and newer releases.
+///
+/// The specified function will be called by the reset handler *after* RAM has been initialized. In
+/// the case of the `thumbv7em-none-eabihf` target the FPU will also be enabled before the function
```
[the remark this answers is not in the document]
Comment (Disasm) [reply macros/src/lib.rs]: Indeed, thanks! Will fix it.

**Pull request merged**